tikv/raftstore/snap.go

```diff
@@ -0,0 +1,834 @@
+	if err != nil {
+		return err
+	}
+	s.MetaFile.File = file
+	s.holdTmpFiles = true
+	for _, cfFile := range s.CFFiles {
+		file, err = os.OpenFile(cfFile.TmpPath, os.O_CREATE|os.O_WRONLY, 0600)
+		if err != nil {
+			return err
+		}
+		if plainFileUsed(cfFile.CF) {
+			cfFile.File = file
+		} else {
+			opts := rocksdb.NewDefaultBlockBasedTableOptions(bytes.Compare)
+			cfFile.SstWriter = rocksdb.NewSstFileWriter(file, opts)
+		}
+	}
+	return nil
+}
+
+func plainFileUsed(cf string) bool {
+	return cf == CFLock
+}
+
+func (s *Snap) readSnapshotMeta() (*rspb.SnapshotMeta, error) {
+	fi, err := os.Stat(s.MetaFile.Path)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	file, err := os.Open(s.MetaFile.Path)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	size := fi.Size()
+	buf := make([]byte, size)
+	_, err = io.ReadFull(file, buf)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	snapshotMeta := new(rspb.SnapshotMeta)
+	err = snapshotMeta.Unmarshal(buf)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	return snapshotMeta, nil
+}
+
+func (s *Snap) setSnapshotMeta(snapshotMeta *rspb.SnapshotMeta) error {
+	if len(snapshotMeta.CfFiles) != len(s.CFFiles) {
+		return errors.Errorf("invalid CF number of snapshot meta, expect %d, got %d",
+			len(s.CFFiles), len(snapshotMeta.CfFiles))
+	}
+	for i, cfFile := range s.CFFiles {
+		meta := snapshotMeta.CfFiles[i]
+		if meta.Cf != cfFile.CF {
+			return errors.Errorf("invalid %d CF in snapshot meta, expect %s, got %s", i, cfFile.CF, meta.Cf)
+		}
+		if util.FileExists(cfFile.Path) {
+			// Check only the file size for `exists()` to work correctly.
+			err := checkFileSize(cfFile.Path, meta.GetSize_())
+			if err != nil {
+				return err
+			}
+		}
+		cfFile.Size = uint64(meta.GetSize_())
+		cfFile.Checksum = meta.GetChecksum()
+	}
+	s.MetaFile.Meta = snapshotMeta
+	return nil
+}
+
+func (s *Snap) loadSnapMeta() error {
+	snapshotMeta, err := s.readSnapshotMeta()
+	if err != nil {
+		return err
+	}
+	err = s.setSnapshotMeta(snapshotMeta)
+	if err != nil {
+		return err
+	}
+	// check if there is a data corruption when the meta file exists
+	// but cf files are deleted.
+	if !s.Exists() {
+		return errors.Errorf("snapshot %s is corrupted, some cf file is missing", s.Path())
+	}
+	return nil
+}
+
+func getDisplayPath(dir string, prefix string) string {
+	cfNames := "(" + strings.Join(snapshotCFs, "|") + ")"
+	return fmt.Sprintf("%s/%s_%s%s", dir, prefix, cfNames, sstFileSuffix)
+}
+
+func (s *Snap) validate(db *badger.DB) error {
+	for _, cfFile := range s.CFFiles {
+		if cfFile.Size == 0 {
+			// Skip empty file. The checksum of this cf file should be 0 and
+			// this is checked when loading the snapshot meta.
+			continue
+		}
+		if plainFileUsed(cfFile.CF) {
+			err := checkFileSizeAndChecksum(cfFile.Path, cfFile.Size, cfFile.Checksum)
+			if err != nil {
+				return err
+			}
+		} else {
+			// TODO: prepare and validate for ingestion
```

Comment:
Use sstReader or badger support Ingestion?

Reply:
It will be supported later.